cuda/btx_cudainterval_callbacks.cpp

```diff
@@ -0,0 +1,589 @@
+  state->context_manager.stream_create_exit({hostname, vpid, vtid}, cuResult, stream);
+}
+// ===============================
+// END Context management
+// ===============================
+
+// ===============================
+// Kernel name and device tracking
+// ===============================
+void module_get_function_entry_callback(void *btx_handle, void *usr_data,
+                                        int64_t ts, const char *event_class_name,
+                                        const char *hostname, int64_t vpid,
+                                        uint64_t vtid, char* name) {
+  std::string name_str(name);
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t hpt = {hostname, vpid, vtid};
+  state->entry_kernel_name[hpt] = name_str;
+}
+
+void module_get_function_exit_callback(void *btx_handle, void *usr_data,
+                                       int64_t ts, const char *event_class_name,
+                                       const char *hostname, int64_t vpid,
+                                       uint64_t vtid, CUresult cuResult,
+                                       CUfunction cuFunction) {
+  if (cuResultIsError(cuResult)) {
+    return;
+  }
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t hpt = {hostname, vpid, vtid};
+  try {
+    auto kernel_name = state->entry_kernel_name.at(hpt);
+    hp_kernel_t hp_kernel_key = {hostname, vpid, cuFunction};
+    state->hp_kernel_to_name[hp_kernel_key] = kernel_name;
+  } catch(const std::out_of_range& oor) {
+    THAPI_FATAL_HPT(hpt,
+                    "entry kernel name not found in module get function exit callback");
+  }
+}
+
+void operation_entry_helper(void *btx_handle, void *usr_data,
+                            int64_t ts, const char *event_class_name,
+                            const char *hostname, int64_t vpid,
+                            uint64_t vtid, CUstream cuStream, CUfunction f) {
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t hpt = {hostname, vpid, vtid};
+  CUdevice dev;
+  bool found = state->context_manager.get_stream_device(hpt, cuStream, &dev);
+  if (!found) {
+    THAPI_FATAL_HPT(hpt, "no device for thread in device op");
+  }
+
+  std::string name = strip_event_class_name(event_class_name);
+  if (f == nullptr) {
+    // for host launch and non launch operations like cuMemcpy, the device op name
+    // is just the function name
+    state->hpt_function_name_to_dev[{hostname, vpid, vtid, name}] = dev;
+    state->hpt_profiled_function_name_and_ts[hpt] = fn_ts_t(name, ts);
+  } else {
+    // for device launch, get the name saved when the kernel was loaded in the
+    // module get function callbacks
+    try
+    {
+      std::string name = state->hp_kernel_to_name.at({hostname, vpid, f});
+      state->hpt_function_name_to_dev[{hostname, vpid, vtid, name}] = dev;
+      state->hpt_profiled_function_name_and_ts[hpt] = fn_ts_t(name, ts);
+    } catch(const std::out_of_range& oor) {
+      THAPI_FATAL_HPT(hpt,
```

Comment:
The only difference between the 2 code path seem to be how we can the name. 

Please make it more obvious. Aka a `if / else` to get the name, then the `hpt_function_name_to_dev` and `hpt_profiled_function_name_and_ts` outside the block